src/ui/EditorNS/editor.cpp

```diff
@@ -152,43 +151,65 @@ namespace EditorNS
-        QString id = data.value("id").toString();
-        QString lang = data.value("lang").toMap().value("name").toString();
-        emit currentLanguageChanged(id, lang);
-    }
-
     void Editor::on_proxyMessageReceived(QString msg, QVariant data)
     {
         emit messageReceived(msg, data);
 
         if(msg == "J_EVT_READY") {
             m_loaded = true;
             emit editorReady();
-        } else if(msg == "J_EVT_CONTENT_CHANGED") {
-            emit contentChanged();
+        } else if(msg == "J_EVT_DOCUMENT_CHANGED") {
+            emit generateChangeActivitySignal(data.toMap());
         } else if(msg == "J_EVT_CLEAN_CHANGED") {
-            emit cleanChanged(data.toBool());
+            //Cache clean state since it's a high-traffic value.
+            m_clean = data.toBool();
+            emit cleanChanged(m_clean);
         } else if(msg == "J_EVT_CURSOR_ACTIVITY") {
-            emit cursorActivity();
+            generateCursorActivitySignal(data.toMap());
         } else if(msg == "J_EVT_GOT_FOCUS") {
             emit gotFocus();
         } else if(msg == "J_EVT_CURRENT_LANGUAGE_CHANGED") {
-            QVariantMap lang = data.toMap();
-            QString id = lang.value("id").toString();
-            QString name = lang.value("lang").toMap().value("name").toString();
-            if (!m_customIndentationMode) {
-                setIndentationMode(id);
-            }
-            emit currentLanguageChanged(id, name);
+            generateLanguageChangeSignal(data.toMap());
         } else if(msg == "J_EVT_DOCUMENT_LOADED") {
             emit documentLoaded(m_alreadyLoaded);
             m_alreadyLoaded = true;
         }
     }
 
+    void Editor::generateChangeActivitySignal(const QVariantMap& v)
+    {
+        UiChangeInfo info;
+        info.charCount = v["charCount"].toInt();
+        info.lineCount = v["lineCount"].toInt();
+
+        emit documentChanged(info);
+    }
+
+    void Editor::generateLanguageChangeSignal(const QVariantMap& v)
+    {
+        QString id = v.value("id").toString();
+        QString name = v.value("lang").toMap().value("name").toString();
+        if (!m_customIndentationMode) {
+            setIndentationMode(id);
+        }
+        emit currentLanguageChanged(id, name);
+    }
+
+    void Editor::generateCursorActivitySignal(const QVariantMap& v)
+    {
+        //NOTE should we cache these values for getCursor, getCharCount
+        //     and getLineCount?  I don't think they're used outside of
+        //     refreshEditorCursorUiInfo.
+        UiCursorInfo info;
+//        info.charCount = v["charCount"].toInt();
+//        info.lineCount = v["lineCount"].toInt();
+        info.cursorLine = v["cursorLine"].toInt();
+        info.cursorColumn = v["cursorColumn"].toInt();
+        info.selectionCharCount = v["selectionCharCount"].toInt();
+        info.selectionLineCount = v["selectionLineCount"].toInt();
+        emit cursorActivity(info);
```

Comment:
So, nothing would speak against saving a copy of `UiCursorInfo info` as a member variable and simply returning it via a `UiCursorInfo Editor::getCursorInfo()`, right? We're doing it wit `isClean` right now, and it sounds like a great idea to have the best of both worlds: non-blocking getters that return data synchronously.